File: lib/autoc/intrusive_hash_set.rb
# frozen_string_literal: true


require 'autoc/std'
require 'autoc/set'
require 'autoc/ranges'


module AutoC


  using STD::Coercions


  # @abstract
  class IntrusiveHashSet < Set

    def range = @range ||= Range.new(self, visibility: visibility)

    attr_reader :load_factor

    attr_reader :_slot, :_slot_p

    def initialize(*args, load_factor: 0.75, auxillaries: false, **kws)
      super(*args, **kws)
      @_slot = decorate(:_slot, abbreviate: true)
      @_slot_p = _slot.lvalue
      @load_factor = load_factor
      @auxillaries = auxillaries
    end

    def render_interface(stream)
      stream << %{
        /** @private */
        typedef struct {
          #{element} element;
        } #{_slot};
      }
      if public?
        stream << %{
          /**
            #{defgroup}

            @brief Unordered collection of unique elements of type #{element}

            This implementation employs open hashing technique.

            For iteration over the set elements refer to @ref #{range}.

            @see C++ [std::unordered_set<T>](https://en.cppreference.com/w/cpp/container/unordered_set)

            @since 2.1
          */
          /**
            #{ingroup}
            @brief Opaque structure holding state of the hash set
            @since 2.1
          */
        }
      else
        stream << PRIVATE
      end
      stream << %{
        typedef struct {
          #{_slot_p} slots; /**< @private */
          size_t size; /**< @private */
          size_t capacity; /**< @private */
        } #{signature};
      }
    end

    def _element = element

  private

    def configure
      super
      method(:void, :tag_empty, { slot: _slot_p }, visibility: :internal)
      method(:int, :is_empty, { slot: _slot_p }, visibility: :internal)
      method(:void, :tag_deleted, { slot: _slot_p }, visibility: :internal)
      method(:int, :is_deleted, { slot: _slot_p }, visibility: :internal)
      method(:int, :tagged, { slot: _slot_p }, visibility: :internal).inline_code %{return #{is_empty}(slot) || #{is_deleted}(slot);}
      method(:void, :create_capacity, { target: lvalue, capacity: :size_t.rvalue }).configure do
        code %{
          size_t slot;
          unsigned char bits = 0;
          assert(target);
          /* fix capacity to become the ceiling to the nearest power of two */
          if(capacity % 2 == 0) --capacity;
          while(capacity >>= 1) ++bits;
          capacity = (size_t)1 << (bits+1); assert(capacity > 0);
          target->size = 0;
          target->capacity = capacity; /* fast slot location for value: hash_code(value) & (capacity-1) */
          target->slots = (#{_slot_p})#{memory.allocate(_slot, :capacity)}; assert(target->slots);
          for(slot = 0; slot < target->capacity; ++slot) #{tag_empty}(target->slots + slot);
        }
        header %{
          @brief Create set with specified capacity

          @param[out] target set to create
          @param[in] capacity initial capacity of the set

          This function creates a new set which should be suffice to contain specific number of elements.

          While the input capacity may be arbitrary the resulting one will be rounded up to the nearest power of two.

          This function may be useful when the (approximate) number of elements this set is about to contain is known in advance
          in order to avoid expensive storage reallocation & elements rehashing operations during the set's lifetime.

          @since 2.1
        }
      end
      method(:size_t, :first_slot, { target: const_rvalue, slot: _slot_p }, visibility: :internal).configure do
        inline_code %{
          assert(target);
          return #{_element.hash_code.('slot->element')} & (target->capacity-1);
        }
      end
      method(:size_t, :next_slot, { target: const_rvalue, slot: :size_t}, visibility: :internal).configure do
        inline_code %{
          assert(target);
          return (slot+1) & (target->capacity-1); /* linear probing */
        }
      end
      method(_slot_p, :_lookup, { target: const_rvalue, value: element.const_rvalue }, visibility: :private).configure do
        code %{
          size_t slot;
          #{_slot_p} next_slot;
          #{_slot} slot_ = { #{value.to_value_argument} };
          assert(target);
          slot = #{first_slot.(target, :slot_)};
          /* zero state signifies real value, deleted state means slot gets skipped, empty state means end of search */
          while(!#{is_empty}(next_slot = target->slots + slot)) {
            if(!#{is_deleted}(next_slot) && #{element.equal.('next_slot->element', value)}) return next_slot;
            slot = #{next_slot}(target, slot);
          }
          return NULL;
        }
      end
      method(:void, :adopt_slot, { target: rvalue, new_slot: _slot_p }, visibility: :internal).configure do
        code %{
          size_t slot;
          #{_slot_p} next_slot;
          assert(target);
          assert(target->size <= target->capacity);
          slot = #{first_slot.(target, new_slot)};
          /* looking for the first slot that is marked either empty or deleted */
          while(!#{tagged}(next_slot = target->slots + slot)) slot = #{next_slot}(target, slot);
          *next_slot = *new_slot;
        }
      end
      method(:void, :_expand, { target: lvalue, force: :int.const_rvalue }, visibility: :private).configure do
        code %{
          assert(target);
          assert(target->capacity*#{load_factor} <= target->capacity); /* guarding againt accidental shrinking */
          if(force || target->size > target->capacity*#{load_factor}) {
            size_t slot;
            size_t source_capacity = target->capacity, source_size = target->size;
            #{_slot_p} source_slots = target->slots;
            #{create_capacity}(target, source_capacity << 1);
            for(slot = 0; slot < source_capacity; ++slot) {
              #{_slot_p} next_slot = source_slots + slot;
              if(!#{tagged}(next_slot)) #{adopt_slot.(target, '*next_slot')};
            }
            target->size = source_size; /* restore the size since create_capacity() resets it to zero */
            #{memory.free(:source_slots)};
          }
        }
      end
      method(:void, :put_force, { target: rvalue, value: element.const_rvalue }, visibility: :internal).configure do
        code %{
          #{_slot} slot;
          assert(target);
          #{element.copy.('slot.element', value)}; /* make a copy right away since adopt() won't do this by itself */
          #{adopt_slot.(target, :slot)};
          ++target->size;
          #{_expand}(target, 0);
        }
      end
      put.configure do
        code %{
          #ifndef NDEBUG
            #{_slot} slot_ = { #{value.to_value_argument} };
          #endif
          assert(target);
          assert(!#{tagged}(&slot_));
          if(!#{_lookup}(target, value)) {
            #{put_force}(target, value);
            return 1;
          } else return 0;
        }
      end
      push.configure do
        code %{
          #{_slot_p} slot;
          #ifndef NDEBUG
            #{_slot} slot_ = { #{value.to_value_argument} };
          #endif
          assert(target);
          assert(!#{tagged}(&slot_));
          if(slot = #{_lookup}(target, value)) {
            #{element.destroy.('slot->element') if element.destructible?};
            #{element.copy.('slot->element', value)};
            return 1;
          } else {
            #{put_force}(target, value);
            return 0;
          }
        }
      end
      remove.configure do
        code %{
          #{_slot_p} slot;
          assert(target);
          if(slot = #{_lookup}(target, value)) {
            #{element.destroy.('slot->element') if element.destructible?};
            #{tag_deleted}(slot);
            --target->size;
            return 1;
          } else return 0;
        }
      end
      default_create.configure do
        dependencies << create_capacity
        inline_code %{
          #{create_capacity.(target, 8)};
        }
      end
      find_first.configure do
        code %{
          #{_slot_p} slot = #{_lookup}(target, value);
          return slot ? &slot->element : NULL;
        }
      end
      copy.configure do
        code %{
          #{range} r;
          assert(target);
          assert(source);
          #{create_capacity}(target, source->capacity);
          for(r = #{range.new}(source); !#{range.empty}(&r); #{range.pop_front}(&r)) #{put}(target, *#{range.view_front}(&r));
        }
      end
      empty.configure do
        inline_code %{
          assert(target);
          return target->size == 0;
        }
      end
      size.configure do
        inline_code %{
          assert(target);
          return target->size;
        }
      end
      contains.configure do
        code %{
          assert(target);
          return #{find_first}(target, value) != NULL;
        }
      end
      destroy_slots_code = %{
        {
          #{range} r;
          for(r = #{range.new}(target); !#{range.empty}(&r); #{range.pop_front}(&r)) {
            #{element.lvalue} e = (#{element.lvalue})#{range.view_front}(&r);
            #{element.destroy.('*e')};
          }
        }
      } if element.destructible?
      destroy.configure do
        code %{
          assert(target);
          #{destroy_slots_code}
          #{memory.free('target->slots')};
        }
      end
      hash_code.configure do
        code %{
          #{range} r;
          size_t hash; /* default incremental hasher is applicable to ordered collections only */
          for(hash = AUTOC_SEED, r = #{range.new}(target); !#{range.empty}(&r); #{range.pop_front}(&r)) {
            #{element.const_lvalue} e = #{range.view_front.(:r)};
            hash ^= #{element.hash_code.('*e')};
          }
          return hash;
        }
      end
      # Return number of equality test operations performed to find element or -1 on failure
      # NOTE this method must stay in sync with #lookup
      method(:int, :count_eops, { target: const_rvalue, value: element.const_lvalue}, constraint:-> { @auxillaries }, visibility: :internal).configure do
        code %{
          int ops = 1;
          size_t slot;
          #{_slot_p} next_slot;
          #{_slot} slot_ = { #{value.to_value_argument} };
          assert(target);
          slot = #{first_slot.(target, :slot_)};
          /* zero state signifies real value, deleted state means slot gets skipped, empty state means end of search */
          while(!#{is_empty}(next_slot = target->slots + slot)) {
            if(!#{is_deleted}(next_slot) && #{element.equal.('next_slot->element', value.to_value_argument)}) return ops;
            slot = #{next_slot}(target, slot);
            ++ops;
          }
          return -1;
        }
      end
      method(:void, :print_stats, { target: const_rvalue, stream: 'FILE*' }, constraint:-> { @auxillaries }, visibility: :private).configure do
        dependencies << AutoC::STD::STDIO_H
        code %{
          int eops = 0, max_eops = 0;
          #{range} r;
          assert(target);
          assert(stream);
          fprintf(stream, "#{type}<#{element}> (#{type.class}<#{element.class}>) @%p\\n", target);
          fprintf(stream, "\\tsize = %zd elements\\n", target->size);
          fprintf(stream, "\\tcapacity = %zd slots\\n", target->capacity);
          fprintf(stream, "\\tslots utilization = %.02f%%\\n", 100.0*target->size/target->capacity);
          fprintf(stream, "\\tbuilt in load factor = #{load_factor}\\n");
          for(r = #{range.new}(target); !#{range.empty}(&r); #{range.pop_front}(&r)) {
            #{element.const_lvalue} x = #{range.view_front}(&r);
            int e = #{count_eops.(target, '*x')}; assert(e >= 0);
            if(max_eops < e) max_eops = e;
            eops += e;
          }
          fprintf(stream, "\\taverage lookup complexity = %.02f equality tests\\n", (double)eops/target->size);
          fprintf(stream, "\\tmaximum lookup complexity = %d equality tests\\n", max_eops);
          fprintf(stream, "\\tlookup compexity histogram:\\n");
          unsigned* eop_count = (unsigned*)calloc(max_eops+1, sizeof(unsigned)); assert(eop_count);
          for(r = #{range.new}(target); !#{range.empty}(&r); #{range.pop_front}(&r)) {
            #{element.const_lvalue} x = #{range.view_front}(&r);
            int e = #{count_eops.(target, '*x')}; assert(e >= 0);
            ++eop_count[e];
          }
          for(int i = 1; i <= max_eops; ++i) {
            fprintf(stream, "\\t\\t%d equality test(s) - for %d or %.02f%% of elements\\n", i, eop_count[i], 100.0*eop_count[i]/target->size);
          }
          free(eop_count);        }
      end
    end

  end # IntrusiveHashSet


  class IntrusiveHashSet::Range < ForwardRange

    def render_interface(stream)
      if public?
        render_type_description(stream)
        stream << %{
          /**
            #{ingroup}
            @brief Opaque structure holding state of the hash set's range
            @since 2.1
          */
        }
      else
        stream << PRIVATE
      end
      stream << %{
        typedef struct {
          #{iterable._slot_p} slots; /**< @private */
          size_t capacity; /**< @private */
          ptrdiff_t slot; /**< @private */
        } #{signature};
      }
    end

  private

    def configure
      super
      custom_create.configure do
        code %{
          assert(range);
          assert(iterable);
          range->slots = iterable->slots;
          range->capacity = iterable->capacity;
          range->slot = -1;
          #{pop_front}(range);
        }
      end
      empty.configure do
        code %{
          assert(range);
          return range->slot >= range->capacity;
        }
      end
      pop_front.configure do
        dependencies << empty
        code %{
          do ++range->slot; while(!#{empty}(range) && #{iterable.tagged}(range->slots + range->slot));
        }
      end
      view_front.configure do
        dependencies << empty
        code %{
          assert(!#{empty}(range));
          return &(range->slots + range->slot)->element;
        }
      end
    end

  end # Range


end

File: src/autoc/intrusive_hash_set.py
import autoc.std as std
from autoc.hash import Xor
from autoc.range import Forward
from autoc.core import _StructRenderer
from autoc.set import Set, _ceil_power2
from autoc.collection import Range as _Range
from autoc.core import out, inout, Macro, Indirection, Callable


class _Macro(Macro):

  def __init__(self, result, parameters, emitter, **kws):
    # Wrap the passthough arguments in () to circumvent operation proirity issues for user-supplied code
    super().__init__(result, parameters, lambda *args: emitter(*(f"({x})" for x in args)), **kws)


#
class Set(_StructRenderer, Set):
  
  def __init__(self, *args, capacity_threshold=0.75, hasher=Xor(), dependencies=(), is_empty, is_deleted, mark_empty, mark_deleted, **kws):
    super().__init__(*args, hasher=hasher, dependencies=(*dependencies, _ceil_power2), **kws)
    self._element_p = Indirection(self.element)
    self.capacity_threshold = capacity_threshold
    self.is_empty = _Macro("int", {"entry": self.element}, is_empty)
    self.is_deleted = _Macro("int", {"entry": self.element}, is_deleted)
    self.mark_empty = _Macro(None, {"entry": out(self.element)}, mark_empty)
    self.mark_deleted = _Macro(None, {"entry": out(self.element)}, mark_deleted)
    self.range = Range(self)

  @property
  def orderable(self):
    return False

  def __setup__(self):
    super().__setup__()
    
    target_i = self.element.variable("target->elements[index]")
    source_i = self.element.variable("source->elements[index]")
    target_elements = self._element_p.variable("target->elements")
    
    with self.size as f:
      f.inline_code = f"""
      assert(target);
      return target->size;
    """
    
    with self.method("int", ("is", "element"), {"element": self.element}, visibility="internal", hidden=True) as f:
      f.code = f"""
        return !({self.is_empty(f.element)} || {self.is_deleted(f.element)});
      """
    
    with self.method(None, "allocate", {"target": inout(self), "capacity": std.size_t}, hidden=True, visibility="internal") as f:
      f.code = f"""
        assert(target);
        assert(capacity > 0);
        target->size = 0;
        target->capacity = _autoc_ceil_power2(capacity);
        assert(!(target->capacity & (target->capacity - 1))); /* verify the capacity is a power of 2 */
        {target_elements} = {self.memory.allocate(self.element, "target->capacity")}; assert({target_elements});
      """
    
    _target = self.variable("_target")
    
    with self.method(None, ("create", "capacity"), {"target": out(self), "capacity": std.size_t}, hidden=True, visibility="internal") as f:
      f.code = f"""
        size_t index;
        assert(target);
        if(capacity) {{
          {self.allocate(f.target, f.capacity)};
          for(index = 0; index < target->capacity; ++index) {self.mark_empty(target_i)};
        }} else {{
          {self.create(f.target)};
        }}
      """

    with self.method(None, ("create", "size"), {"target": out(self), "size": std.size_t}) as f:
      f.code = f"""
        assert(target);
        {self.create_capacity(f.target, f"(size_t)({f.size}/{self.capacity_threshold})")};
      """

    with self.create as f:
      f.inline_code = f"""
        assert(target);
        target->elements = NULL;
        target->capacity = target->size = 0;
      """
    
    with self.method(Callable.Parameter(self._element_p), ("locate", "element"), {"target": self, "_index": out(std.size_t), "element": self.element}, visibility="internal", hidden=True, constraint=lambda: self.element.comparable) as f:
      f.code = f"""
        size_t index, start;
        {self._element_p} _element = NULL;
        assert(target);
        assert(_index);
        assert({self.is_element(f.element)});
        /* linear probing */
        if(target->elements) {{
          assert(target->capacity > 0);
          start = {self.element.hash_lookup_hash(f.element)} & (target->capacity-1); /* capacity is assumed to be the power of 2 */
          /* lookup terminator for the existing entry is an empty slot while deleted slot is not */
          for(index = start; index < target->capacity; ++index) {{
            if(!({self.is_empty(target_i)})) {{
              if(!({self.is_deleted(target_i)}) && {self.element.hash_lookup_equal(target_i, f.element)}) {{
                _element = &{target_i};
                goto stop;
              }}
            }} else goto stop;
          }}
          for(index = 0; index < start; ++index) {{
            if(!({self.is_empty(target_i)})) {{
              if(!({self.is_deleted(target_i)}) && {self.element.hash_lookup_equal(target_i, f.element)}) {{
                _element = &{target_i};
                goto stop;
              }}
            }} else goto stop;
          }}
          stop:
            *_index = index;
            return _element;
        }}
        return NULL;
      """
    
    with self.method(Callable.Parameter(self._element_p), ("locate", "slot"), {"target": self, "_index": out(std.size_t), "element": self.element}, visibility="internal", hidden=True, constraint=lambda: self.element.comparable) as f:
      f.code = f"""
        size_t index, start;
        assert(target);
        assert(_index);
        assert(target->capacity > 0);
        assert(target->size < target->capacity);
        assert({self.is_element(f.element)});
        /* linear probing */
        start = {self.element.hash_lookup_hash(f.element)} & (target->capacity-1); /* capacity is assumed to be the power of 2 */
        /* lookup terminator for non-existing entry is either empty or deleted slot */
        for(index = start; index < target->capacity; ++index) {{
          if(!{self.is_element(target_i)}) {{
            *_index = index;
            return &{target_i};
          }}
        }}
        for(index = 0; index < start; ++index) {{
          if(!{self.is_element(target_i)}) {{
            *_index = index;
            return &{target_i};
          }}
        }}
        abort(); /* not finding a suitable empty slot is a fatal error */
      """

    with self.method(None, "resize", {"target": inout(self), "new_size": std.size_t}, hidden=True, constraint=lambda: self.element.copyable) as f:
      f.code = f"""
        {_target.definition};
        size_t index, _index, new_capacity;
        assert(target);
        new_capacity = _autoc_ceil_power2((size_t)(new_size/{self.capacity_threshold})); /* predict new capacity after size changing respecting the desired capacity threshold */
        if(new_capacity < target->capacity && new_capacity < (size_t)(target->size/{self.capacity_threshold})) new_capacity = target->capacity; /* prevent the new capacity to fall below minimum required for original set's elements */
        if(new_capacity < 8) new_capacity = 8; /* enforce minimum viable capacity */
        assert(target->size <= new_capacity); /* require the new capacity to be large enough for the set to contain all original elements */
        if(new_capacity != target->capacity) {{
          {self.create_capacity(_target, "new_capacity")};
          if(target->elements) {{
            for(index = 0; index < target->capacity; ++index) {{
              if({self.is_element(target_i)}) {{
                {self.element.copy(self.locate_slot(_target, "&_index", target_i), target_i)}; /* direct planting in order to prevent from triggering the subsequent resizings */
              }}
            }}
            _target.size = target->size;
            {self.destroy(f.target)};
          }}
          *target = _target;
        }}
      """
    
    with self.contains as f:
      f.code = f"""
        size_t index;
        assert(target);
        assert({self.is_element(f.element)});
        return target->elements && {self.locate_element(f.target, "&index", f.element)} != NULL;
      """
    
    with self.destroy as f:
      if self.element.destructible:
        f.code = f"""
          size_t index;
          assert(target);
          if(target->elements) {{
            for(index = 0; index < target->capacity; ++index)
              if({self.is_element(target_i)})
                {self.element.destroy(target_i)};
            {self.memory.free(target_elements)};
          }}
        """
      else:
        f.code = f"""
          assert(target);
          if(target->elements) {{
            {self.memory.free(target_elements)};
          }}
        """

    with self.empty as f:
      f.inline_code = f"""
        assert(target);
        return target->size == 0;
      """
    
    with self.put as f:
      f.code = f"""
        size_t index;
        assert(target);
        assert({self.is_element(f.element)});
        if(!{self.contains(f.target, f.element)}) {{
          {self.resize(f.target, "target->size+1")};
          {self.element.copy(self.locate_slot(f.target, "&index", f.element), f.element)};
          ++target->size;
          return 1;
        }} else return 0;
      """
    
    with self.remove as f:
      _destroy = self.element.destroy(target_i) if self.element.destructible else str()
      f.code = f"""
        size_t index;
        assert(target);
        assert({self.is_element(f.element)});
        if({self.locate_element(f.target, "&index", f.element)}) {{
          {_destroy};
          {self.mark_deleted(target_i)};
          --target->size;
          return 1;
        }} else return 0;
      """
    
    with self.method(self.element.view_type, ("find", "view"), {"target": self, "element": self.element}) as f:
      f.code = f"""
        size_t index;
        assert(target);
        assert({self.is_element(f.element)});
        return {self.locate_element(f.target, "&index", f.element).bind(f.result)};
      """
    
    with self.copy as f:
      f.code = f"""
        size_t index;
        assert(target);
        assert(source);
        {self.create_size(f.target, "source->size")};
        for(index = 0; index < source->capacity; ++index) {{
          if({self.is_element(source_i)}) {{
            {self.put(f.target, source_i)};
          }}
        }}
      """

    range = self.range
    r = range.variable("r")
    
    with self.equal as f:
      f.code = f"""
        {r.definition};
        assert(left);
        assert(right);
        if({self.size(f.left)} == {self.size(f.right)}) {{
          for({r} = {range.new(f.left)}; !{range.empty(r)}; {range.move_front(r)}) {{
            if(!{self.contains(f.right, range.front_view(r))}) return 0;
          }}
          return 1;
        }} else return 0;
      """
    
    state = self.hasher.state_t.variable("state")
    
    with self.hash as f:
      f.code = f"""
        size_t result;
        {r.definition};
        {state.definition};
        assert(target);
        {self.hasher.create(state)};
        for({r} = {range.new(f.target)}; !{range.empty(r)}; {range.move_front(r)}) {{
          {self.hasher.update(state, self.element.hash(range.front_view(r)))};
        }}
        result = {self.hasher.hash(state)};
        {self.hasher.destroy(state)};
        return result;
      """

  def _render_struct(self, stream):
    super()._render_struct(stream)
    if self.public:
      stream.append("/** @public */\n")
    stream.append(f"""typedef struct {{
      {self.element}* elements; /**< @private */
      {std.size_t} capacity; /**< @private */
      {std.size_t} size; /**< @private */
    }} {self.name};
    """)


#
class Range(_Range, Forward):
  
  def render_declarations(self, stream, header):
    super().render_declarations(stream, header)
    if header:
      stream.append(f"""
        typedef struct {{
          {Indirection(self.iterable, constant=True)} iterable; /**< @private */
          {std.size_t} front; /**< @private */
        }} {self.name};
      """)

  def _copy(self, result, parameters, **kws):
    return Macro(result, parameters, lambda target, source: f"{target} = {source}", **kws)

  def __setup__(self):
    super().__setup__()
    
    front_element = self.element.variable("target->iterable->elements[target->front]")

    with self.method(None, "next", {"target": inout(self)}, hidden=True, visibility="internal") as f:
      f.code = lambda: f"""
        assert(target);
        while(!{self.empty("target")} && !{self.iterable.is_element(front_element)}) ++target->front;
      """
    
    with self.method(Callable.Parameter(self), "new", {"iterable" : self.iterable}) as f:
      result = f.result.variable("result")
      f.code = f"""
        {result.definition};
        assert(iterable);
        result.iterable = iterable;
        result.front = 0;
        {self.next("&result")};
        return {result};
      """

    with self.empty as f:
      f.code = f"""
        assert(target);
        return target->front >= target->iterable->capacity;
      """
    
    with self.front as f:
      result = f.result.variable("result")
      f.code = lambda: f"""
        {result.definition};
        assert(target);
        assert(!{self.empty(f.target)});
        assert({self.iterable.is_element(front_element)});
        {self.element.copy(result, front_element)};
        return {result};
      """
        
    with self.front_view as f:
      f.code = lambda: f"""
        assert(target);
        assert(!{self.empty(f.target)});
        assert({self.iterable.is_element(front_element)});
        return {front_element.bind(self.iterable.element.view_type)};
      """
    
    with self.move_front as f:
      f.code = f"""
        assert(target);
        assert(!{self.empty(f.target)});
        ++target->front;
        {self.next(f.target)};
      """    